### PR TITLE
Introduce a settings DSL for central dependency declaration

### PR DESCRIPTION
**Pull request by melix**
This commit introduces a new DSL, available on `Settings` via the
dependency resolution management block, which allows declaring aliases
for dependencies and bundles of dependencies.

```
dependencyResolutionManagement {
    dependenciesModel {
        alias("groovy", "org.codehaus.groovy", "groovy") {
            strictly("3.0.2")
        }
    }
}
```

Doing this, Gradle will automatically generate type-safe accessors
which are exposed as extensions to all projects. This effectively
allows sharing dependency declarations between projects.

Said differently, this is an officially supported pattern for the
various "dependency version sharing" patterns that we found in the
wild:

- declaring dependency versions in gradle.properties, then referencing
the version via `project.someLibraryVersion`
- creating a new class in `buildSrc` which contains dependency coordinates
and then can be accessed in the different projects
- creating extra properties for dependency versions at the root project
- using external plugins like the "refresh dependencies" plugin
- ...

In addition to this DSL, we introduce a TOML file, which, if found under
the `gradle` directory, will be used to source dependency aliases.

For example, giving the following file:

```
[dependencies]
guava = { group = "com.google.guava", name = "guava", version = "27.0-jre" }
groovy = { group = "org.codehaus.groovy", name = "groovy", version.strictly = "3.0.2" }
```

a dependency can be added in a project using the type-safe `libs` extension:

```
dependencies {
    api(libs.guava)
}
```

This file also supports bundles of dependencies, in case more than one
dependency needs to be added:

```
[bundles]
groovy = ["groovy", "groovy-json"]
```

then dependencies can be added via:

```
dependencies {
   implementation(libs.groovyBundle)
}
```

The file is configuration-cache safe: any change to the file will invalidate the
cache. But more importantly, it's build cache safe: if the aliases don't change,
there's no need to rebuild the dependent scripts. This means that changing this
file by changing, for example, the dependency versions, will not trigger a
recompilation of build scripts like some of the approaches described above.

The TOML file also lets you declare plugin versions:

```
[plugins]
my.awesome.plugin="1.0.2"
```

which allows application of the plugin without version in build scripts:

```
plugins {
   id 'my.awesome.plugin'
}
```

In addition, we also generate type-safe accessors for projects. For example:

```
dependencies {
    api(project(":commons:core"))
}
```

can be replaced with:

```
dependencies {
    api(projects.commons.core)
}
```

Therefore any change to project coordinates, or removals of subprojects would
trigger a build script compilation error, avoiding tedious search and replace.

Fixes #?

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by vlsi** [comment]
> In addition, we also generate type-safe accessors for projects. For example:

That is awesome.

> In addition to this DSL, we introduce a TOML file, which, if found under the gradle directory

Do you think it could support overriding the version with a command-line flag?

What is the approach for [file splitting](https://github.com/toml-lang/toml/issues/397)?
What is the approach for plugin versions?

**Comment by melix** [comment]
> Do you think it could support overriding the version with a command-line flag?

Technically speaking we can. Now I'd have to talk to the larger team to say if we _should_ :) Note that if you override a dependency version, it would probably only override the default model, so if you have something like:

```
dependencies {
    implementation(libs.guava) {
        version {
           // this wins over whatever is overridden
        }
    }
}
```

> What is the approach for file splitting?

For now I'd say KISS. We could provide alternatives if there's a legitimate interest in making this more complicated. Alternatively settings plugins can provide this feature. Last but not least, because we do code generation we could hit the limit of the number of methods in a class...

> What is the approach for plugin versions?

This PR lets you declare this:

```
[plugins]
my.awesome.plugin="1.4"
```

then you don't have to specify a version when applying in build scripts.

**Comment by vlsi** [comment]
> Technically speaking we can. Now I'd have to talk to the larger team to say if we should :)

Let me clarify the use-case.

I want to verify if my project works with a newer version of a third-party dependency.
For instance, I might need to verify if a newer Checkstyle version works for my project, or a newer Guava works for me.
The important point is I don't want to modify the source files because, well, editing the source files is prone to errors.

Note: I can't use `composite builds` for that, because:
1) Third-party is not always Gradle based (yet :'( )
2) `composite build` differs from the case when dependency is fetched from Maven repository. For instance, composite build ignores Gradle Metatada, so the only way to verify if Gradle Metatada is OK is to install it to a repository (e.g. `~/.m2`) and build against it.

Here's the case where I install `Apache Calcite Avatica` to the local maven repository and build `Apache Calcite` with that dependency: https://github.com/apache/calcite/blob/cd922deff8ae3f25546b1d77fb147c3098eb177b/.github/workflows/main.yml#L88-L103
I use a hand-crafted version like `1.0.0-dev-master-SNAPSHOT` to ensure it does not accidentally fetch an old snapshot from somewhere.

Here's the case when Checkstyle CI job verifies the new Checkstyle works for verifying of pgjdbc sources: https://github.com/checkstyle/checkstyle/blob/3c2249b239cbfe002af3649f1f7c1f9ae61df3b1/.ci/wercker.sh#L60-L71

**Comment by vlsi** [comment]
> Alternatively settings plugins can provide this feature

I wonder if TOML could be separate from the code generator, so out-of-core plugins could leverage the same generator.

**Comment by melix** [comment]
> I wonder if TOML could be separate from the code generator, so out-of-core plugins could leverage the same generator.

That's exactly the case: the TOML file is just, if you will, a standard plugin hooking into the code generator. In the end, there's a single code generator and settings plugins can contribute entries. But in the end we're going to be limited by the number of methods which can be defined in a class file.

**Comment by melix** [comment]
> Let me clarify the use-case.

That is certainly a legitimate use case.

**Comment by martinbonnin** [comment]
That looks seriously awesome 🤩 . Thanks for addressing this !! 

Will the TOML file support dependencies "groups" that need to be all the same version so that it can be changed in a single place? For an example that obviously won't work since I don't think TOML supports variable but just to demonstrate the goal: 

```toml
[versions]
sqldelight = "1.7.0"

[dependencies]
sqlDelightPlugin = { group = "com.squareup.sqldelight", name = "gradle-plugin", version = "${versions.sqldelight}" }
sqlDelightDriverAndroid = { group = "com.squareup.sqldelight", name = "android-driver", version.strictly = "${versions.sqldelight}" }
sqlDelightDriverAndroid = { group = "com.squareup.sqldelight", name = "native-driver", version.strictly = "${versions.sqldelight}" }

```

Also would it be possible to copy/paste the maven coordinates in one go? That's a minor change but not having to separate group/artifactId/version would be a nice addition as most of the documentations use that representation:

```toml
[dependencies]
guava = { gav = "com.google.guava:guava:27.0-jre" }
```

Which leads me to a wider question: did you consider a Kotlin script for this that would expose a strongly type way to register dependencies and could be self documented? Something like `dependencies.gradle.kts`:

```kotlin
dependencies {
  create("guava") {
    group = "com.google.guava"
    artifact = "guava"
    version= "27.0-jre"
  }
  // or the shorthand gav version
  create("guava") {
    gav = "com.google.guava:guava:27.0-jre"
  }
}

bundles {
  create("groovy") {
    from("groovy", "groovy-json")
  }
}
```

Could that work?

**Comment by melix** [comment]
> Will the TOML file support dependencies "groups" that need to be all the same version so that it can be changed in a single place? 

Not out of the box. If we do this I'd rather add a `[versions]` section and have the other modules reference it via a `versionRef="someId"` rather than inventing a substitution mechanism.

> Also would it be possible to copy/paste the maven coordinates in one go? That's a minor change but not having to separate group/artifactId/version would be a nice addition as most of the documentations use that representation:

That's already supported by this PR :)

> Which leads me to a wider question: did you consider a Kotlin script for this that would expose a strongly typed way to register dependencies and could be self documented?

There's already the settings DSL for this. Note, however, that using Kotlin or the settings DSL would invalidate all build scripts classpath and trigger recompilation of all scripts if any version or coordinate changes, whereas using the TOML this would only happen if you add/remove an alias, not if you change GAV coordinates.

**Comment by martinbonnin** [comment]
> Not out of the box. If we do this I'd rather add a [versions] section and have the other modules reference it via a versionRef="someId" rather than inventing a substitution mechanism.

Yup, makes sense 👍 

> That's already supported by this PR :)

Neat ! \o/

> There's already the settings DSL for this.

Cool! Is there an exemple somewhere?

> using Kotlin or the settings DSL would invalidate all build scripts classpath and trigger recompilation of all scripts if any version or coordinate changes

That's a pretty big downside and certainly one of the main reason not a lot of people use this? (I think?)

**Comment by melix** [comment]
> Cool! Is there an exemple somewhere?

It's in the issue description ;)

**Comment by martinbonnin** [comment]
> It's in the issue description ;)

Got it, Thanks 👍  I was looking in the current production docs 🤦 . What about moving that DSL to a separate `dependencies.gradle.kts` file that wouldn't trigger recompilation of all scripts when changed?

**Comment by vlsi** [comment]
>melix: using Kotlin or the settings DSL would invalidate all build scripts classpath and trigger recompilation of all scripts if any version or coordinate changes

AFAIK `settings.gradle.kts` does not serve as a classpath for other scripts, so why recompile?

I just checked a couple of modifications to `settings.gradle.kts`, and file edits do not seem to trigger the recompilation of other `build.gradle.kts` scripts.

I tried to edit a string literal inside `settings.gradle.kts`, and I tried to declare a function `fun hello() = "42"`. The changes went OK, and only `settings` compilation was necessary.

**Comment by melix** [comment]
no you're right, using the settings API this would be fine, I was mistaken with the current buildSrc approach.

**Comment by melix** [comment]
> I was looking in the current production docs

There are no docs because we didn't decide to go with this proposal yet.

**Comment by martinbonnin** [comment]
This PR got me thinking... Would it make sense for the TOML file to be generic and not only about dependencies? A kind of supercharged type-safe `gradle.properties`? As far as I can tell, people started using `buildSrc` and such other solutions mostly for type safety and autocomplete so I can see them switching to something like:

```toml
# properties.toml
[libs]
guava = "com.google.guava:guava:27.0-jre"
```

```kotlin
// build.gradle.kts
dependencies {
   // can't think of a better name than `typesafeProperties` but something shorter would be nice
   implementation(typesafeProperties.libs.guava)
}
```

Of course this is less expressive than having the TOML file know about dependencies but I _think_ specifying the versions constraints, etc can be left to a `platform` project?

Having a generic `toml` format would allow for other properties to be used in a type safe manner. For an exemple for publishing:

```toml
# properties.toml
[publishing]
url = "https://github.com/me/my-library"
group = "com.example"
version = "1.0"
```

```kotlin
// build.gradle.kts
group = typesafeProperties.publishing.group
version = typesafeProperties.publishing.version
publishing {
    publications {
        create<MavenPublication>("mavenJava") {
            pom {
                name.set("My Library")
                description.set("A concise description of my library")
                url.set(version = typesafeProperties.publishing.url)
            }
        }
    }
}
```

This could also be used to share properties between multiple projects, inject credentials, or whatever people use `gradle.properties` today? I understand that deviates quite a bunch from the original intent of the PR but that could maybe address most of the use cases in a more generic way? Any thoughts?

**Comment by melix** [comment]
@martinbonnin I don't think we should do this. The big advantage of this DSL and TOML file is that we know what the properties are used for. It allows for some optimizations and it allows better modeling. We can discuss its extension to other use cases but we shouldn't mix all things together IMO.

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by gabrielittner** [comment]
Will a plugin or dependency declared in the TOML file also be usable in the buildSrc/build.gradle(.kts) build script? Just asking because buildSrc is special

**Comment by melix** [comment]
> Will a plugin or dependency declared in the TOML file also be usable in the buildSrc/build.gradle(.kts) build script? Just asking because buildSrc is special

The `buildSrc` directory should be considered as an included build. As such, it can have its own settings, own TOML file, etc... So if you want to share versions between the main build and `buildSrc`, the idea would be to have a settings plugin that you apply to both. Ideally we should have a way to do this via an included build so that it can be defined in the same "project", but there are some quirks today (it's a chicken and egg problem) and we're thinking about potential solutions to this.

**Comment by gabrielittner** [comment]
> Ideally we should have a way to do this via an included build so that it can be defined in the same "project"

This would be great. I guess until then it could be possible to just symlink the TOML file for buildSrc to the regular one. That's what we're doing with `buildSrc/gradle.properties` right now.

**Comment by gildor** [comment]
> We can discuss its extension to other use cases but we shouldn't mix all things together IMO.

I think it's a very important use-case for Gradle in general, of course not in the context of this particular pull request, maybe it worth reporting a separate feature request as for supporting built-in Gradle properties or other plugins? 
Typesafe configuration and static config files (with overrides, like we have now for gradle.properties) is a very powerful feature which can improve UX of Gradle a lot

**Comment by melix** [comment]
> Typesafe configuration and static config files (with overrides, like we have now for gradle.properties) is a very powerful feature which can improve UX of Gradle a lot

This is my personal opinion here, but type safe APIs which don't know about what they are talking about is not any more helpful than "dynamic" APIs. Type safe == completion in the IDE here, which is, IMO, not about type safety.

That's what I mean by "we shouldn't mix things together": for every use case, there's a real, latent, type-safe API we can build. For each use case, a different solution. For example this one can handle the dependency notations nicely without recompiling everything because it _knows_ what it's talking about.

**Comment by martinbonnin** [comment]
> For example this one can handle the dependency notations nicely without recompiling everything because it knows what it's talking about.

What is specific about dependencies that allows to skip recompiling the build scripts in that case? 

Since I can change a Gradle property in `gradle.property` without having to recompile my `build.gradle.kts` I was expecting that the same could be done using a `toml` file and generated accessors? As long as the accessors ABI doesn't change, the build scripts compilation could be skipped?

**Comment by melix** [comment]
This is precisely because I can build a model where the generated accessors _do not_ depend on anything else than the aliases being used. There is semantics associated to those aliases, it's not a random string in the properties file. So I know that if we change, say, the GAV coordinates, the accessors don't even have to be rebuilt.

**Comment by melix** [comment]
@bot-gradle test this

**Comment by gildor** [comment]
> Type safe == completion in the IDE here, which is, IMO, not about type safety
Yes, agree with it, but IDE support is also important by itself. But also even basic type safety (string vs object vs int vs non-existent property) is also nice, which I believe can be provided by generating code from toml file, it would be already an improvement

**Comment by gildor** [comment]
I have one more question about semantics of `strictly` versioning, it looks quite similar (at first glance) to java-platform, does it correct? And can it replace it (at least for some basic cases), when you want to declare a strict version of dependency for your entire project, but do not plan to distribute own BOM dependency

**Comment by melix** [comment]
> I have one more question about semantics of strictly versioning, it looks quite similar (at first glance) to java-platform, does it correct? And can it replace it (at least for some basic cases), when you want to declare a strict version of dependency for your entire project, but do not plan to distribute own BOM dependency

No, that's incorrect. A platform is just a bag of constraints. Those dependency constraints may, or may not, use strict versions. Similarly, here the versions declared in settings or via this TOML file may, or may not, use strict versions. You could use strict versions in this file, but it wouldn't remove the need for platforms, for example, to talk about transitive dependencies you don't directly use.

**Comment by DPUkyle** [comment]
I'm definitely interested in this, if nothing else as a way to define a dependency version once and reuse everywhere in a repo with multiple subprojects.

**Comment by melix** [comment]
@bot-gradle test QF

**Comment by melix** [comment]
@bot-gradle test QF

**Comment by melix** [comment]
@bot-gradle test QF

**Comment by melix** [comment]
@bot-gradle test QF

**Comment by melix** [comment]
@bot-gradle test QF

File: buildSrc/subprojects/jvm/src/main/kotlin/gradlebuild.api-parameter-names-index.gradle.kts [hunk 1]
```diff
@@ -36,6 +36,7 @@ val parameterNamesIndex by tasks.registering(ParameterNamesIndex::class) {
         project.layout.buildDirectory.file(moduleIdentity.baseName.map { "generated-resources/$it-parameter-names/$it-parameter-names.properties" })
     )
 }
+
```
Comment (melix) [inline buildSrc/subprojects/jvm/src/main/kotlin/gradlebuild.api-parameter-names-index.gradle.kts:39]: In case you wonder this is because the cache is busted on CI, this shouldn't be necessary at all.

**Comment by melix** [comment]
@bot-gradle test QF

**Comment by melix** [comment]
@bot-gradle test QF

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by Vampire** [comment]
This sounds nice.
Disclaimer: I only read the description, not the source.
I have three questions that come to my mind when I read this.

1. Is it possible to configure the path for the TOML, or maybe even better include one TOML into the other? When I e. g. have the main build and `buildSrc`, or some included builds instead of `buildSrc`, I might prefer one place to define all versions.

2. Can the TOML be extended to also define arbitrary versions? There are versions that are not dependencies or plugins, but make sense to also maintain at the same place, for example JaCoCo version, CodeNarc version, ...

3. When 2. is possible, an easy utility to access them in the actual "production" code would be nice. So if you are using this in `buildSrc` or a `buildSrc`-like included build, you can easily then use the defined JaCoCo version to configure the JaCoCo plugin in a pre-compiled script plugin and similar things.

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
As discussed with @ljacomet I'm optimistically merging this in order to get more feedback from users. Will do post-merge review updates.

**Comment by JavierSegoviaCordoba** [comment]
@melix is it possible to use the accessors inside `buildSrc`? I am trying to use the libs accessor inside a precompiled plugin but it is an unresolved reference.

**Comment by melix** [comment]
> is it possible to use the accessors inside buildSrc? I am trying to use the libs accessor inside a precompiled plugin but it is an unresolved reference.

You can't use the generated accessors in a precompiled script plugin for now because they don't belong to the same projects. The accessors "belong to" the project they are applied to. You can see `buildSrc` as an included build, so an independent one. The precompiled plugins _could_ be applied to the main build, but then the libs that you see there come from a different context. While we could do something special for `buildSrc`, this is not the road we're taking now, which is to try to make `buildSrc` different from any other included build.

So, basically, you can reuse the accessors between the `buildSrc` and main build _build scripts_, but not within the "production code" of the plugins themselves, because you don't know in what context they are going to be used.

**Comment by JavierSegoviaCordoba** [comment]
Can I see the plans for `buildSrc` somewhere?

**Comment by melix** [comment]
There are currently no plans to improve `buildSrc` in this context. Doesn't mean there won't be, just that we need to settle things down a bit before.

**Comment by Vampire** [comment]
What about my comment above? :-)
Or was it misplaced here?

**Comment by melix** [comment]
Sorry I missed it, is was misplaced... between other comments :D

> Is it possible to configure the path for the TOML, or maybe even better include one TOML into the other? When I e. g. have the main build and buildSrc, or some included builds instead of buildSrc, I might prefer one place to define all versions.

You can import from an arbitrary location :

```
dependencyResolutionManagement {
    dependenciesModel("testLibs") {
        from(files("gradle/test-dependencies.toml"))
    }
}
```
However the TOML file itself doesn't allow importing other files.

> Can the TOML be extended to also define arbitrary versions? There are versions that are not dependencies or plugins, but make sense to also maintain at the same place, for example JaCoCo version, CodeNarc version, ...

Both the DSL and the TOML file allow declaring arbitrary versions:

```
[versions]
jacoco=0.8.6
groovy=3.0.5

[dependencies]
groovy = { module="org.codehaus.groovy:groovy", version.ref="groovy" }
groovy-json = { module="org.codehaus.groovy:groovy-json", version.ref="groovy" }
```

The versions can be changed when "importing" the model. Versions are rich:

```
dependenciesModel {
     version("my-lib") {
         strictly("[1.0, 2.0]")
         require("1.5")
     }
}
```
but for convenience in build scripts `myLibVersion` would only return the _required_ version as this is the most common use case.

> When 2. is possible, an easy utility to access them in the actual "production" code would be nice. So if you are using this in buildSrc or a buildSrc-like included build, you can easily then use the defined JaCoCo version to configure the JaCoCo plugin in a pre-compiled script plugin and similar things.

As I mentioned in the comment above, this is not possible yet. There's a context problem here. Basically the dependency model is available to build scripts and that's it. Not to plugins or Java code, because we don't know in which context it is used so the class we would inject doesn't reflect what you would get at runtime.

**Comment by gabrielittner** [comment]
> As I mentioned in the comment above, this is not possible yet. There's a context problem here. Basically the dependency model is available to build scripts and that's it. Not to plugins or Java code, because we don't know in which context it is used so the class we would inject doesn't reflect what you would get at runtime.

Do you have any suggestions on how to deal with dependencies that are declared in precompiled script plugins? Since the generated accessors are not available you'd end up duplicating dependency declarations again if a precompiled script plugin wants to declare dependencies.

**Comment by Vampire** [comment]
> > Is it possible to configure the path for the TOML, or maybe even better include one TOML into the other? When I e. g. have the main build and buildSrc, or some included builds instead of buildSrc, I might prefer one place to define all versions.
> 
> You can import from an arbitrary location :
> 
> ```
> dependencyResolutionManagement {
>     dependenciesModel("testLibs") {
>         from(files("gradle/test-dependencies.toml"))
>     }
> }
> ```
> 
> However the TOML file itself doesn't allow importing other files.

Ah, nice, that will do I think.
Especially as you can also provide multilple files if I interpret the example correctly.
I guess the overwrite behaviour if a version is in multiple files is defined and will be documented?

> > When 2. is possible, an easy utility to access them in the actual "production" code would be nice. So if you are using this in buildSrc or a buildSrc-like included build, you can easily then use the defined JaCoCo version to configure the JaCoCo plugin in a pre-compiled script plugin and similar things.
> 
> As I mentioned in the comment above, this is not possible yet. There's a context problem here. Basically the dependency model is available to build scripts and that's it. Not to plugins or Java code, because we don't know in which context it is used so the class we would inject doesn't reflect what you would get at runtime.

I'm not sure I expressed my expectation correctly, so here a little example:
- main project build script applies `com.example.foo`
- `build-src` build has the TOML file including `jacoco = "1.2.3"`
- `build-src/src/main/kotlin/com/example/foo.gradle.kts` now should be able to use `jacocoVersion` for doing `jacoco { toolVersion = whatever.jacocoVersion }`

This should always result in JaCoCo 1.2.3 being configured, no matter what context.
Maybe one could generate a class including those versions for usage within the production project.
I guess it could now be done manually by having another build included by `build-src`, that references the same TOML, iterates over the versions, generates a class and then compiles that class.
But it would probably be nice to have this (or something similar) conveniently provided by Gradle. :-)

The same of course for the actual dependencies, to declare dependencies in pre-compiled script plugins like @gabrielittner said.

**Comment by melix** [comment]
> I'm not sure I expressed my expectation correctly, so here a little example:

No, that's exactly about this. Currently the versions are not exposed to precompiled script plugins (or production code in `buildSrc`). One reason is that if we use the same mechanism, say, that you define a dependencies model in `buildSrc` and that you make it available in the precompiled script plugins, then a plugin could have:

```
jacoco {
    toolVersion = libs.jacocoVersion.get()
}
```

The `libs` that you see here is a generated class which "belongs to" `buildSrc` code. It has a package name and class name which are generated for the `buildSrc` context. However, in a precompiled script plugin _execution phase_, the context is the project itself.

So the `libs` object that you would see would _at best_ be the same because you imported the same model in both main build and buildSrc, but it could be totally different. Worse there are some name clashes.

So sharing with production code, because of class shadowing, is a bit tricky. We can certainly use a different namespace to make it explicit, maybe use a different prefix to access the extensions, ... But I think we should take time to think about the solution and not rush here.

**Comment by tprochazka** [comment]
I really want some official solution for this dependency mess.
I was thinking that the platform module is the official way, until now.
I'm quite confused that it is completely independent and partially redundant to platform modules.
I know that the platform module is just about versions, but it can be extended to generating aliases or not?
And I'm even more confused with creating a new file format TOML, instead of a primary focus on Kotlin DSL with existing IDE support. There is DSL support but it looks weird, but maybe the example in the first post just doesn't show everything. I would expect something more similar to this TOML file.

**Comment by melix** [comment]
> I was thinking that the platform module is the official way, until now.
> I'm quite confused that it is completely independent and partially redundant to platform modules.

It's expected because there is an overlap between the two, and you might want to use both at the same time. A platform is a component in a dependency graph, which can be upgraded on its own (for example if you have the Spring BOM in a dependency graph, it can be upgraded to a different version because of transitives), participates into dependency resolution (constraints declared in a platform will apply even if you don't use them), is inherited by consumers (a platform is a node in the graph so consumers will see it). Anything declared in a platform will influence any graph which uses this platform, including transitive dependencies. A catalog, like proposed here, is very lean: it’s only used for first level dependencies, in order to pick versions. It has no influence on the consumer and any “unused” dependency in a catalog will not influence the resolution. The consumer of a catalog is Gradle itself, while the consumer of a platform is any library, Gradle, Maven, etc.

This proposal is about fixing the "redundant declaration" use case, which isn't really addressed today and that people implement in various ways.

>  I know that the platform module is just about versions, but it can be extended to generating aliases or not?

This [follow-up PR](https://github.com/gradle/gradle/pull/15019) actually does something like that: it lets you declare a version catalog from a platform definition. The choice of how to consume the platform, either as a platform, or as a catalog, or both, is a consumer choice.

Catalogs can also be composed from different sources easily, without "leaking" to consumers, which is one of the drawbacks of platforms today.

**Comment by tprochazka** [comment]
I started to use it and it works great, in TOML file is possible to generate dependencies in three different ways

```
groovy = { group = "org.codehaus.groovy", name = "groovy", version = "3.0.2" }
groovy1 = { module = "org.codehaus.groovy:groovy", version = "3.0.2" }
groovy2 = "org.codehaus.groovy:groovy:3.0.2"
```
The disadvantages of TOML are missing code completion and error messages which not only tell the line with error but it also not tell that this error is related to TOML file parsing until you use --stacktrace

DSL currently allows just this
```
dependencyResolutionManagement {
    dependenciesModel("libs") {
        alias("framework-core").to("com.company.android:android-framework-core:1.5.5")
        alias("framework-core").to("com.company.android","android-framework-core").version("1.5.5")
    }
}
There is no way how to use `org.codehaus.groovy:groovy` as in TOML file

If you want to create a bundle from multiple library with the same version it is possible in this way via DSL

```
dependencyResolutionManagement {
    dependenciesModel("libs") {
        version("lifecycle", "1.5.5")
        alias("androixLifecycleRuntime").to("androidx.lifecycle","lifecycle-runtime-ktx").versionRef("lifecycle")
        alias("androixLifecycleViewmodel").to("androidx.lifecycle","lifecycle-viewmodel-ktx").versionRef("lifecycle")
        bundle("androixLifecycle", listOf("androixLifecycleRuntime", "androixLifecycleViewmodel"))
    }
}

I would maybe expect some more structuralized way like 

```
        bundle("androixLifecycle").version("1.5.5") {
            alias("androixLifecycleRuntime").to("androidx.lifecycle","lifecycle-runtime-ktx")
            alias("androixLifecycleViewmodel").to("androidx.lifecycle","lifecycle-viewmodel-ktx")
            alias("someDifferentLibInBundle").to("xx","yy").version("1.0.0")
        }
```
or just without bundle
```
        version("1.5.5") {
            alias("androixLifecycleRuntime").to("androidx.lifecycle","lifecycle-runtime-ktx")
            alias("androixLifecycleViewmodel").to("androidx.lifecycle","lifecycle-viewmodel-ktx")
        }
```
or 
```
       bundle("basicLibs") {
           version("1.5.5") {
               alias("androixLifecycleRuntime").to("androidx.lifecycle","lifecycle-runtime-ktx")
               alias("androixLifecycleViewmodel").to("androidx.lifecycle","lifecycle-viewmodel-ktx")
           }
           alias("someDifferentLibInBundle").to("xx","yy").version("1.0.0")
       }
```
It is visually much more clear that it is a group that belong together.

I would prefer DSL way over TOML, but it is quite verbose and it must be inside of settings.gradle.kts file, separated will would be better.

And I miss a possibility to create a more structuralized libs hierarchy by using "." inside of alias name, which will create a new packages inside of libs object like

`libs.androidx.<library>`

**Comment by tprochazka** [comment]
And what about maven dependencies with classifier? It looks that it is completely unsupported :-(

**Comment by gildor** [comment]
I'm playing with new versionCatalog DSL, works great, but I found big general issue for my common use case: convention plugins.

So if I have dependencies.toml, I can use it any any module and in buildSrc (if include it), but I cannot use dependencies versions in my precompiled convention plugin: `buildSrc/main/kotlin/some-convention.gradle.kts`

I understand why it happening, but I think it's a big limitation in general, because I apply many dependencies using different convention plugins, so even if I able to read toml file, I will not get any generated accessor.

Should it be possible somehow to generate libs accessors also for precompiled script plugins (which already support Kotlin DSL accessors generation)?

**Comment by melix** [comment]
We have ideas how to make this possible, but probably not for the first release.

**Comment by gabrielittner** [comment]
@melix Is there an issue to follow for this?

**Comment by ljacomet** [comment]
> And what about maven dependencies with classifier? It looks that it is completely unsupported :-(

The dependency catalog focuses on dependency coordinates, namely the Group, Artifact and Version.
We discussed the idea of including classifiers or even variant information and decided against it. Indeed these have a context linked to their application and thus it is better expressed in the build file rather than in the catalog.

**Comment by tprochazka** [comment]
Do you know why this doesn't work for me in gradle-6.8-milestone-3?

```
dependencies {
    api(projects.commons.core)
}
```
 It was removed meantime?

**Comment by melix** [comment]
Yes, this was removed from 6.8. You can use `master` instead, this is still under development.

**Comment by tprochazka** [comment]
Thanks! So it will be in 6.9, right? I will keep it for now without it, I want to put into production with 6.8, 6.9 is too far.

**Comment by tprochazka** [comment]
There are more cases when it fails like this one

```
val ccaApi by configurations
ccaApi(libs.uiSkeleton) {
    because("Some comment why why need this dependency")
}

```
ends with

`e: ...lib_ui_skeleton\build.gradle.kts:34:12: Type mismatch: inferred type is Provider<MinimalExternalModuleDependency!>! but String was expected
`

**Comment by melix** [comment]
> There are some cases when it fails like this one

Can you please file issues for those cases?

**Comment by tprochazka** [comment]
It looks that in gradle-6.8-rc-1 it completely stops working :-( 
I was expecting that only propejct() support was removed.
@melix If you mean to report it here https://discuss.gradle.org/c/bugs/11 I have disabled New Topic button there.

**Comment by melix** [comment]
> I was expecting that only propejct() support was removed.

No the whole central dependencies declaration has been removed.

And I mean creating issues with the reproducers for what you've found. This ticket is closed, there have been many follow ups and if you don't want your bug reports to be lost it's better to track them with proper issues.

**Comment by JavierSegoviaCordoba** [comment]
@melix should be great there is a place where check all info about this feature. I always visit this issue.

**Comment by tprochazka** [comment]
@melix Done. I hope that I did it in the right way.  ;-)

**Comment by tprochazka** [comment]
> @melix should be great there is a place where check all info about this feature. I always visit this issue.

Yes. At least slack channel for this would be great ;-)

**Comment by tprochazka** [comment]
I just did a script that can generate a TOML file for a multimodule project, including versions ref extraction, and replace all old dependencies with libs.xyz in the next step. Except for lack of IDE support, it works great. I also already implemented it on my project, because I was expected that it will be in 6.8 :-(  At least as an experimental feature.

**Comment by melix** [comment]
Please refer to this issue for tracking progress: https://github.com/gradle/gradle/issues/15352

**Comment by tprochazka** [comment]
Weird that I can see, that this #15352 issue has "0 of 12 tasks complete", but I'm unable to find which tasks they are :-(

**Comment by gildor** [comment]
@tprochazka It's just about checklists in PR description (see `Contributor Checklist` and `Gradle Core Team Checklist` in this PR description)

**Comment by melix** [comment]
@tprochazka Yeah I noticed that. We're using ZenHub to track epics and unfortunately you can't see anything. This is what you should see:

image
